test/functional/HaddockComments.hs

```diff
@@ -0,0 +1,69 @@
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE ViewPatterns #-}
+
+module HaddockComments
+  ( tests,
+  )
+where
+
+import Control.Monad.IO.Class (liftIO)
+import Data.Foldable (find)
+import Data.Maybe (mapMaybe)
+import Data.Text (Text)
+import qualified Data.Text.IO as T
+import Language.Haskell.LSP.Test
+import Language.Haskell.LSP.Types
+import System.FilePath ((<.>), (</>))
+import Test.Hls.Util
+import Test.Tasty
+import Test.Tasty.HUnit
+
+tests :: TestTree
+tests =
+  testGroup
+    "haddock comments"
+    [ normal "HigherRankFunction.hs" Signature 4 6,
+      normal "KindSigFunction.hs" Signature 9 10,
+      normal "MultivariateFunction.hs" Signature 2 8,
+      normal "QualFunction.hs" Signature 2 10,
+      normal "Record.hs" Record 7 2,
+      expectedNothing "StaleFunction.hs" Signature 3 3,
+      expectedNothing "StaleRecord.hs" Record 3 12
+    ]
+
+normal :: FilePath -> GenCommentsType -> Int -> Int -> TestTree
+normal fp (toTitle -> expectedTitle) l c = testCase fp $
+  runSession hlsCommand fullCaps haddockCommentsPath $ do
+    doc <- openDoc fp "haskell"
+    _ <- waitForDiagnostics
+    actions <- getCodeActions doc (Range (Position l c) (Position l $ succ c))
+    case find ((== Just expectedTitle) . caTitle) actions of
+      Just (CACodeAction x) -> do
+        executeCodeAction x
+        contentAfterAction <- documentContents doc
+        expected <- liftIO . T.readFile $ haddockCommentsPath </> fp <.> "expected"
```

Comment:
BTW, would you mind if I ask you to change the order of extensions?
I think it will be easier to check test results in an editor or browser if there is some syntax highlighting.
I mean, something like `StaleRecord.expected.hs`